e3nn_jax/_src/so3grid.py

```diff
@@ -0,0 +1,169 @@
+    @property
+    def res_beta(self) -> int:
+        return self.s2_signals.res_beta
+
+    @property
+    def res_alpha(self) -> int:
+        return self.s2_signals.res_alpha
+
+    @property
+    def res_theta(self) -> int:
+        return self.s2_signals.shape[-3]
+
+    @property
+    def grid_theta(self) -> jnp.ndarray:
+        return jnp.linspace(0, 2 * jnp.pi, self.res_theta)
+
+    @property
+    def grid_resolution(self) -> str:
+        return (self.res_theta, self.res_beta, self.res_alpha)
+
+    @staticmethod
+    def from_function(
+        func: Callable[[jax.Array], float],
+        res_beta: int,
+        res_alpha: int,
+        res_theta: int,
+        quadrature: str,
+        *,
+        dtype: jnp.dtype = jnp.float32,
+    ) -> "SO3Signal":
+        """Create a signal on the sphere from a function of rotations.
+
+        Args:
+            func (`Callable`): function on the sphere that maps a 3x3 rotation matrix to a scalar or vector.
+            res_theta: resolution for theta (for the angle in the axis-angle parametrization)
+            res_beta: resolution for beta (for the axis in the axis-angle parametrization)
+            res_alpha: resolution for alpha (for the axis in the axis-angle parametrization)
+            quadrature: quadrature to use
+            dtype: dtype of the signal
+
+        Returns:
+            `SO3Signal`: signal on SO(3)
+        """
+        # Construct the grid over S2 for the axis in the axis-angle parametrization.
+        # The shape of s2_signals will be different at the end,
+        # but we just extract the axes (via grid_vectors) from it now.
+        s2_signals = e3nn.SphericalSignal.zeros(
+            res_beta, res_alpha, quadrature=quadrature, dtype=dtype
+        )
+        axes = s2_signals.grid_vectors
+        angles = jnp.linspace(0, 2 * jnp.pi, res_theta)
+
+        assert axes.shape == (res_beta, res_alpha, 3)
+        assert angles.shape == (res_theta,)
+
+        # Construct the rotation matrices for each (axis, angle) pair.
+        Rs_fn = e3nn.axis_angle_to_matrix
+        Rs_fn = jax.vmap(Rs_fn, in_axes=(0, None))
+        Rs_fn = jax.vmap(Rs_fn, in_axes=(None, 0))
+        Rs = Rs_fn(axes, angles)
+        assert Rs.shape == (res_theta, res_beta, res_alpha, 3, 3)
+
+        # Evaluate the function at each rotation matrix.
+        func_fn = jax.vmap(jax.vmap(jax.vmap(func)))
+        fs = func_fn(Rs)
+
+        # Move the function output axes to the front.
+        if fs.ndim > 3:
+            for _ in range(fs.ndim - 3):
+                fs = jnp.moveaxis(fs, -1, 0)
+
+        batch_dims = fs.shape[:-3]
+        assert fs.shape == (*batch_dims, res_theta, res_beta, res_alpha)
+
+        # Account for angle-dependency in Haar measure.
+        fs = fs * (1 - jnp.cos(angles))[..., None, None]
+        s2_signals = s2_signals.replace_values(fs)
+        assert s2_signals.shape == (*batch_dims, res_theta, res_beta, res_alpha)
+        return SO3Signal(s2_signals)
```

Comment:
Seems weird to me that the `SO3Signal` does not contain the actual values of the function but weighted values. Is there a way to change that?

Reply:
yes, this indeed is a bug which I fixed in https://github.com/e3nn/e3nn-jax/pull/87